source/extensions/upstreams/http/udp/upstream_request.cc

```diff
@@ -0,0 +1,141 @@
+      dispatcher_(dispatcher) {
+  socket_->ioHandle().initializeFileEvent(
+      dispatcher_, [this](uint32_t) { onSocketReadReady(); }, Event::PlatformDefaultTriggerType,
+      Event::FileReadyType::Read);
+}
+
+void UdpUpstream::encodeData(Buffer::Instance& data, bool /*end_stream*/) {
+  for (const Buffer::RawSlice& slice : data.getRawSlices()) {
+    absl::string_view mem_slice(static_cast<const char*>(slice.mem_), slice.len_);
+    if (!capsule_parser_.IngestCapsuleFragment(mem_slice)) {
+      ENVOY_LOG_MISC(error, "Capsule ingestion error occured: slice = {}", mem_slice);
+      break;
+    }
+  }
+  capsule_parser_.ErrorIfThereIsRemainingBufferedData();
+}
+
+Envoy::Http::Status UdpUpstream::encodeHeaders(const Envoy::Http::RequestHeaderMap&,
+                                               bool /*end_stream*/) {
+  Api::SysCallIntResult rc = socket_->connect(host_->address());
+  if (SOCKET_FAILURE(rc.return_value_)) {
+    return absl::InternalError("Upstream socket connect failure.");
+  }
+  // Synthesize the 200 response headers downstream to complete the CONNECT-UDP handshake.
+  Envoy::Http::ResponseHeaderMapPtr headers{
+      Envoy::Http::createHeaderMap<Envoy::Http::ResponseHeaderMapImpl>(
+          {{Envoy::Http::Headers::get().Status, "200"},
+           {Envoy::Http::Headers::get().CapsuleProtocol, "?1"}})};
+  upstream_to_downstream_->decodeHeaders(std::move(headers), false);
+  return Envoy::Http::okStatus();
+}
+
+void UdpUpstream::resetStream() { upstream_to_downstream_ = nullptr; }
+
+void UdpUpstream::onSocketReadReady() {
+  uint32_t packets_dropped = 0;
+  const Api::IoErrorPtr result = Network::Utility::readPacketsFromSocket(
+      socket_->ioHandle(), *socket_->connectionInfoProvider().localAddress(), *this,
+      dispatcher_.timeSource(), /*prefer_gro=*/false, packets_dropped);
+  if (result == nullptr) {
+    socket_->ioHandle().activateFileEvents(Event::FileReadyType::Read);
+    return;
+  }
+}
+
+void UdpUpstream::processPacket(Network::Address::InstanceConstSharedPtr /*local_address*/,
+                                Network::Address::InstanceConstSharedPtr /*peer_address*/,
+                                Buffer::InstancePtr buffer, MonotonicTime /*receive_time*/) {
+  std::string data = buffer->toString();
+  quiche::QuicheBuffer serialized_capsule =
+      SerializeCapsule(quiche::Capsule::Datagram(data), &capsule_buffer_allocator_);
+
+  Buffer::InstancePtr capsule_data = std::make_unique<Buffer::OwnedImpl>();
+  capsule_data->add(serialized_capsule.AsStringView());
+  bytes_meter_->addWireBytesReceived(capsule_data->length());
+  upstream_to_downstream_->decodeData(*capsule_data, false);
+}
+
+bool UdpUpstream::OnCapsule(const quiche::Capsule& capsule) {
+  quiche::CapsuleType capsule_type = capsule.capsule_type();
+  if (capsule_type != quiche::CapsuleType::DATAGRAM) {
+    // Silently drops Datagram Capsules with an unknown type.
```

Comment:
nit: these aren't datagram capsules
```suggestion
    // Silently drops capsules with an unknown type.
```

Reply:
Can we get datagram capsules support?  this is critical to the implementation for Zero Trust Access use cases.